includes/class-wc-checkout.php

```diff
@@ -197,555 +290,505 @@ public function create_order() {
+		WC()->session->set( 'chosen_shipping_methods', $chosen_shipping_methods );
+		WC()->session->set( 'chosen_payment_method', $data['payment_method'] );
 
-					if ( is_wp_error( $new_customer ) ) {
-						throw new Exception( $new_customer->get_error_message() );
-					} else {
-						$this->customer_id = absint( $new_customer );
-					}
+		// Update cart totals now we have customer address.
+		WC()->cart->calculate_totals();
+	}
 
-					wc_set_customer_auth_cookie( $this->customer_id );
 
-					// As we are now logged in, checkout will need to refresh to show logged in data
-					WC()->session->set( 'reload_checkout', true );
+	/**
+	 * Process an order that does require payment.
+	 *
+	 * @since  2.7.0
+	 * @param  int $order_id
+	 * @param  string $payment_method
+	 */
+	protected function process_payment( $order_id, $payment_method ) {
+		$available_gateways = WC()->payment_gateways->get_available_payment_gateways();
 
-					// Also, recalculate cart totals to reveal any role-based discounts that were unavailable before registering
-					WC()->cart->calculate_totals();
+		if ( ! isset( $available_gateways[ $payment_method ] ) ) {
+			return;
+		}
 
-					// Add customer info from other billing fields
-					if ( $this->posted['billing_first_name'] && apply_filters( 'woocommerce_checkout_update_customer_data', true, $this ) ) {
-						$userdata = array(
-							'ID'           => $this->customer_id,
-							'first_name'   => $this->posted['billing_first_name'] ? $this->posted['billing_first_name'] : '',
-							'last_name'    => $this->posted['billing_last_name'] ? $this->posted['billing_last_name'] : '',
-							'display_name' => $this->posted['billing_first_name'] ? $this->posted['billing_first_name'] : '',
-						);
-						wp_update_user( apply_filters( 'woocommerce_checkout_customer_userdata', $userdata, $this ) );
-					}
-				}
+		// Store Order ID in session so it can be re-used after payment failure
+		WC()->session->set( 'order_awaiting_payment', $order_id );
 
-				// Do a final stock check at this point
-				$this->check_cart_items();
+		// Process Payment
+		$result = $available_gateways[ $payment_method ]->process_payment( $order_id );
 
-				// Abort if errors are present
-				if ( wc_notice_count( 'error' ) > 0 ) {
-					return false;
-				}
+		// Redirect to success/confirmation/payment page
+		if ( isset( $result['result'] ) && 'success' === $result['result'] ) {
+			$result = apply_filters( 'woocommerce_payment_successful_result', $result, $order_id );
 
-				$order_id = $this->create_order();
+			if ( is_ajax() ) {
+				wp_send_json( $result );
+			} else {
+				wp_redirect( $result['redirect'] );
+				exit;
+			}
+		}
+	}
 
-				if ( is_wp_error( $order_id ) ) {
-					throw new Exception( $order_id->get_error_message() );
-				}
+	/**
+	 * Process an order that doesn't require payment.
+	 *
+	 * @since  2.7.0
+	 * @param  int $order_id
+	 */
+	protected function payment_complete( $order_id ) {
```

Comment:
`payment_complete` seems like an odd name for a function that process an order that doesn't require payment. Maybe we should call it something else?